File: generated_code/GeminiPro/Java/019_SplayTree.java
/**
 * A complete, self-contained Java class implementing a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again. This implementation
 * provides insert, delete, and search operations.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        /**
         * Constructor for a new node.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Default constructor to create an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splay operation. It moves the node with the given key to the root of the tree.
     * If the key is not found, the last accessed node is moved to the root.
     * This implementation uses a top-down splaying approach.
     * @param key The key to splay on.
     */
    private void splay(int key) {
        if (root == null) {
            return;
        }

        // Use a dummy node to simplify linking.
        // The left subtree of the dummy will hold the "right tree" (nodes > key).
        // The right subtree of the dummy will hold the "left tree" (nodes < key).
        Node dummy = new Node(0);
        Node leftTreeMax = dummy;
        Node rightTreeMin = dummy;
        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) {
                    break;
                }
                // Zig-Zig case (left-left)
                if (key < current.left.key) {
                    current = rightRotate(current);
                    if (current.left == null) {
                        break;
                    }
                }
                // Link the current root to the right tree
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) {
                    break;
                }
                // Zig-Zig case (right-right)
                if (key > current.right.key) {
                    current = leftRotate(current);
                    if (current.right == null) {
                        break;
                    }
                }
                // Link the current root to the left tree
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else {
                // Key found
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;
        rightTreeMin.left = current.right;
        current.left = dummy.right;
        current.right = dummy.left;
        root = current;
    }

    /**
     * Searches for a key in the tree.
     * This operation performs a splay on the accessed key, moving it (or the last
     * node on the search path) to the root.
     * @param key The key to search for.
     * @return The Integer key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        splay(key);
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree.
     * If the key already exists, the tree is splayed on that key but not modified.
     * After a new insertion, the new node becomes the root of the tree.
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, insert the new key as the root.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key.
        // If the key is present, it becomes the root.
        // If not, the node that would be its parent becomes the root.
        splay(key);

        // If the key is already in the tree, do nothing.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
            root = newNode;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
            root = newNode;
        }
    }

    /**
     * Deletes a key from the Splay Tree.
     * If the key is found, it is splayed to the root, removed, and the remaining
     * subtrees are joined.
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the key to the root.
        splay(key);

        // If the key is not in the tree, do nothing.
        if (root.key != key) {
            return;
        }

        // Now the node to be deleted is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right child becomes the new root.
            root = rightSubtree;
        } else {
            // Make the left subtree the new tree.
            root = leftSubtree;
            // Splay on the key `key` in the new tree. Since `key` is larger than any
            // element in this subtree, this will bring the maximum element of the
            // left subtree to the root.
            splay(key);
            // Attach the original right subtree to the new root.
            root.right = rightSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/019_SplayTree.py
class SplayTree:
    """
    A complete, self-contained Python class that implements a Splay Tree.

    This class provides a dictionary-like set for storing unique integers.
    The key feature of a Splay Tree is that recently accessed elements are
    quick to access again. It achieves this by moving any accessed (searched,
    inserted, or deleted) node to the root of the tree through a series of
    rotations, an operation known as "splaying".

    The main public methods are:
    - insert(key): Adds a key to the tree.
    - delete(key): Removes a key from the tree.
    - search(key): Checks for a key and splays the tree.

    Attributes:
        root (SplayTree._Node): The root node of the tree, or None if empty.
    """

    class _Node:
        """A private inner class representing a node in the Splay Tree."""
        def __init__(self, key):
            self.key = key
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _splay(self, key):
        """
        Performs the top-down splay operation.

        The node with the given key (or the last node accessed if the key is
        not found) is moved to the root of the tree. This is a private helper
        method that forms the core of the Splay Tree's functionality.

        Args:
            key: The key to splay the tree around.
        """
        if not self.root:
            return

        # Dummy node simplifies linking. Its right child will form the left
        # splay tree, and its left child will form the right splay tree.
        dummy = self._Node(None)
        left_tree_tail = dummy
        right_tree_tail = dummy
        current = self.root

        while True:
            if key < current.key:
                if not current.left:
                    break
                if key < current.left.key:  # Zig-Zig (Right-Right rotation)
                    temp = current.left
                    current.left = temp.right
                    temp.right = current
                    current = temp
                    if not current.left:
                        break
                # Link to the right tree (for nodes larger than key)
                right_tree_tail.left = current
                right_tree_tail = current
                current = current.left
            elif key > current.key:
                if not current.right:
                    break
                if key > current.right.key:  # Zag-Zag (Left-Left rotation)
                    temp = current.right
                    current.right = temp.left
                    temp.left = current
                    current = temp
                    if not current.right:
                        break
                # Link to the left tree (for nodes smaller than key)
                left_tree_tail.right = current
                left_tree_tail = current
                current = current.right
            else:  # key is found
                break

        # Reassemble the tree
        left_tree_tail.right = current.left
        right_tree_tail.left = current.right
        current.left = dummy.right
        current.right = dummy.left
        self.root = current

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the tree is splayed around that key,
        but no new node is added. Otherwise, a new node with the key is
        created and inserted, becoming the new root after splaying.

        Args:
            key (int): The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        self._splay(key)

        # If key is already in the tree after splaying, do nothing
        if self.root.key == key:
            return

        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            self.root.left = None
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            self.root.right = None
        self.root = new_node

    def search(self, key):
        """
        Searches for a key in the tree and performs the splaying operation.

        This method splays the tree on the accessed node if the key is found,
        or on its would-be parent if the key is not found. This moves the
        relevant node to the root.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False
        self._splay(key)
        return self.root.key == key

    def delete(self, key):
        """
        Deletes a key from the tree.

        If the key is found, it is splayed to the root and then removed.
        The remaining subtrees are joined. If the key is not found, the tree
        is splayed around the last accessed node, but no deletion occurs.

        Args:
            key (int): The integer key to delete.
        """
        if not self.root:
            return

        self._splay(key)

        # If key is not in the tree after splaying, do nothing
        if self.root.key != key:
            return

        # Key is at the root, now perform deletion
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If no left child, the right subtree becomes the new tree
            self.root = right_subtree
        else:
            # Splay the left subtree for the key. This brings the largest
            # element (the predecessor of the deleted key) to the root of
            # the left subtree. This new root will have no right child.
            self.root = left_subtree
            self._splay(key)
            
            # Attach the original right subtree
            self.root.right = right_subtree

    def __str__(self):
        """Returns an in-order string representation of the tree's keys."""
        if not self.root:
            return "SplayTree()"
        keys = self._in_order_traversal(self.root)
        return f"SplayTree({keys})"

    def _in_order_traversal(self, node):
        """Helper for recursive in-order traversal to get keys as a list."""
        if not node:
            return []
        return (self._in_order_traversal(node.left) +
                [node.key] +
                self._in_order_traversal(node.right))